src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -1129,7 +1129,16 @@ public abstract static class AnyInstanceVariableNode extends PrimitiveArrayArgum
 
         @Specialization(limit = "getDynamicObjectCacheLimit()")
         protected boolean any(RubyDynamicObject self,
-                @CachedLibrary("self") DynamicObjectLibrary objectLibrary) {
+                @CachedLibrary("self") DynamicObjectLibrary objectLibrary,
+                @Cached BranchProfile hasPropertiesProfile) {
+            var shape = objectLibrary.getShape(self);
+
+            if (shape.getPropertyCount() == 0) {
```

Comment:
This count doesn't include hidden properties.
This is fine since anyway we only consider String keys, not `HiddenKey`, but might deserve a comment.

Reply:
Hmm... I thought this went the other way. I was worried that it would include hidden properties so the quick check would fail if hidden properties exist but String properties don't. `ShapeImpl#getKeyArray` allocates its array size based on `getPropertyCount`.

Reply:
Right, so both of these methods ignore hidden properties, so it's all fine. The docs are not very explicit about that, but nothing we can do about that in this PR. So fine as-is.